### PR TITLE
[WIP] Update to rotor filter component

### PR DESCRIPTION
**Pull request by pavankum**
Added functionality to filter rotors by minimum no. of rotors to be present, additional `filter_by` arg in the function which defaults to 'max' so that previous tests and functionality are unaffected, and added new tests

## Description
If we decide to fragment molecules during dataset creation that have rotors above a threshold value this filter helps in keeping only molecules that meet this threshold. 

## Todos
  - [x] Filters rotors by minimum number of rotors to be present

## Status
- [ ] Ready to go

### CODE REVIEW

**Comment by jthorton** [comment]
@pavankum Thanks for putting this together this looks great. One thing I think could be changed is the `filter_by` argument this would then let users filter by both max and min at the same time in one go like the element/weight filters. Users could still filter by max/min if we allow the other to take `None` which would skip the check so we would have something like

```python
# filter molecules with between 2 and 6 rotatable bonds
rotor_filter.maximum_rotors = 6 
rotor_filter.minimum_rotors = 2

# filter molecules with at least 2 rotatable bonds
rotor_filter.maximum_rotors = None
rotor_filter.minimum_rotors = 2

# filter molecules with no more than 5 rotatable bonds
rotor_filter.maximum_rotors = 5
rotor_filter.minimum_rotors = None
```
otherwise users wishing to filter by max and min in the workflow would have to add the component twice which is possible but more work.

**Comment by pavankum** [comment]
@jthorton Thanks Josh for the review. Yeah, it makes sense to have a general filter, I made the changes.

File: qcsubmit/workflow_components/filters.py [hunk 1]
```diff
@@ -349,18 +349,20 @@ def provenance(self) -> Dict:
...
         Rotatable bonds are non terminal torsions found using the `find_rotatable_bonds` method of the
         openforcefield.topology.Molecule class.
     """
 
     component_name = "RotorFilter"
-    component_description = (
-        "Filter the molecules based on the maximum number of allowed rotatable bonds."
+    component_description = "Filter the molecules based on the minimum/maximum number of allowed rotatable bonds."
+    component_fail_message = "The molecule has too many(low) rotatable bonds."
+    minimum_rotors: int = Field(
```
Comment (jthorton) [inline qcsubmit/workflow_components/filters.py:362]: This should now have the type `Optional[int]` to show that `None` can be supplied. I think we should also explain what happens if `None` is supplied in the field description.

File: qcsubmit/workflow_components/filters.py [hunk 2]
```diff
@@ -385,9 +387,11 @@ def _apply(self, molecules: List[Molecule]) -> ComponentResult:
 
         # run the the molecules and calculate the number of rotatable bonds
         for molecule in molecules:
-            if len(molecule.find_rotatable_bonds()) > self.maximum_rotors:
+            if (
```
Comment (jthorton) [inline qcsubmit/workflow_components/filters.py:390]: We need some logic to handle the case where `maximum_rotors` or `minimum_rotors` is `None`. Lets also only call `find_rotatable_bonds` once and save the result for the comparison.

**Comment by jthorton** [comment]
@pavankum thanks for making the changes only the last two things to change and this should be good to go.

File: qcsubmit/workflow_components/filters.py (hunk 2 above)
Comment (pavankum) [reply qcsubmit/workflow_components/filters.py]: > @pavankum thanks for making the changes only the last two things to change and this should be good to go.

@JoshHorton My apologies for the delay in fixing these issues. Please review the changes when you have time. Thank you.

**Comment by jthorton** [comment]
@pavankum Thanks for starting this I have finished this in #132 as this has been parked for a while let me know if you find any issues.

**Pull request closed without merging**